Перенесен код из старого проекта

YanikMax opened the pull request.

Commit:
Перенесен код из старого проекта

Commit:
1) Пересоздал проект, сделал первый коммит и перенес в новый файл код до исправления.
2) Кнопки + и datePicker перенес в NavBarItem
3) Добавил возможность скрывать клавиатуру в строке поиска и в строке названия трекера
4) Добавил реализацию скролла таблицы с расписанием на маленьких экранах
5) Добавлен лейбл счетчика, а так же склонение слова "дня"
6) Добавил реализацию запрета отметки карточки трекера для будущих дат
7) Строка поиска теперь работает (реализовано нажатием на Fix)
8) Убраны force unwrapping из кода

Данила, прошу прощения, что некорректно создал репозиторий. Сейчас это новый репозиторий с исправлением тех ошибок, что Вы указали.

DanilaG commented on TrackerYP/TabBar.swift:
```diff
@@ -0,0 +1,35 @@
+        tabBar.layer.borderWidth = 1
+        tabBar.layer.masksToBounds = true
+        
+        let trackersViewController = TrackersViewController()
```
Для того чтобы `TrackersViewController` использовался во `UINavigationController`, его нужно обернуть в данном месте (на этапе добавления в иерархию). Таким образом, должно получится что-то вроде:
`let trackersViewController = UINavigationController(rootViewController: TrackersViewController())`

DanilaG commented on TrackerYP/Tracker/TrackerViewController.swift:
```diff
@@ -105,15 +107,43 @@ final class TrackersViewController: UIViewController {
+    
+    private func configureNavigationBar() {
+        let trackersViewController = TrackersViewController()
+        let navigationController = UINavigationController(rootViewController: trackersViewController)
```
Нет необходимости создовать `navigationController`, который не используется, давай избавимся от него, удалив данные стрки. А в функции будем настраивать уже существующий контролер (т. е. себя)

DanilaG commented on TrackerYP/Tracker/TrackerViewController.swift:
```diff
@@ -105,15 +107,43 @@ final class TrackersViewController: UIViewController {
+        
+        // Создаем экземпляр UIBarButtonItem для datePicker
+        let datePickerButtonItem = UIBarButtonItem(customView: datePicker)
+        trackersViewController.navigationItem.rightBarButtonItem = datePickerButtonItem
```
Давай настраивать тот view controller, в котором мы сейчас. Для этого `trackersViewController` нам не нужен, его создание можно удалить, тогда, в тех местах где он используется мы должны изменить текущий view controller (self). Но так как self можно опустить, то получим замену
`trackersViewController.navigationItem.leftBarButtonItem = addButtonItem` на `navigationItem.leftBarButtonItem = addButtonItem`, и аналогично для правого BarButtonItem

DanilaG commented on TrackerYP/Tracker/TrackerViewController.swift:
```diff
@@ -105,15 +107,43 @@ final class TrackersViewController: UIViewController {
+        let trackersViewController = TrackersViewController()
+        let navigationController = UINavigationController(rootViewController: trackersViewController)
+        
+        // Создаем UIView и добавляем кнопку как его подпредставление
+        let addButtonContainerView = UIView()
+        addButtonContainerView.addSubview(addButton)
```
Нет необходимости дополнительно оборачивать `addButton `)
Можно удалить и использовать в коде ниже `addButton`

DanilaG commented on TrackerYP/Tracker/TrackerViewController.swift:
```diff
@@ -0,0 +1,374 @@
+private extension TrackersViewController {
+    func configureViews() {
+        view.backgroundColor = .white
+        [trackerLabel, addButton, datePicker, searchField, collectionView, mainSpacePlaceholderStack, searchSpacePlaceholderStack].forEach { view.addSubview($0) }
```
Давай удалим `addButton` и `datePicker` из списка на добавление на view нашего view controller. Так как эти элементы мы добавляем в navigation bar. При этом, нужно не забыть удалить констрейнты

DanilaG commented on TrackerYP/Tracker/TrackerViewController.swift:
```diff
@@ -0,0 +1,374 @@
+        view.backgroundColor = .white
+        [trackerLabel, addButton, datePicker, searchField, collectionView, mainSpacePlaceholderStack, searchSpacePlaceholderStack].forEach { view.addSubview($0) }
+        
+        addButton.translatesAutoresizingMaskIntoConstraints = false
```
Давай удалим лишнию настройку `addButton` для view, так как `addButton` теперь будет использоваться в navigation bar.

DanilaG commented on TrackerYP/Tracker/TrackerViewController.swift:
```diff
@@ -0,0 +1,374 @@
+        
+        addButton.translatesAutoresizingMaskIntoConstraints = false
+        trackerLabel.translatesAutoresizingMaskIntoConstraints = false
+        datePicker.translatesAutoresizingMaskIntoConstraints = false
```
Давай удалим лишнию настройку `datePicker ` для view, так как `datePicker` теперь будет использоваться в navigation bar.

DanilaG commented on TrackerYP/Tracker/TrackerViewController.swift:
```diff
@@ -0,0 +1,374 @@
+            trackerLabel.topAnchor.constraint(equalTo: view.topAnchor, constant: view.frame.height * 0.1083),
+            trackerLabel.leftAnchor.constraint(equalTo: view.leftAnchor, constant: 18),
+            
+            datePicker.widthAnchor.constraint(equalToConstant: 120),
+            datePicker.centerYAnchor.constraint(equalTo: addButton.centerYAnchor),
+            datePicker.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor, constant: -16),
+            
+            addButton.leftAnchor.constraint(equalTo: view.leftAnchor, constant: 18),
+            addButton.topAnchor.constraint(equalTo: view.topAnchor, constant: view.frame.height * 0.07019),
```
Давай удалим лишнию настройку `datePicker ` и `addButton` для view, так как `datePicker` и `addButton` теперь будет использоваться в navigation bar.

DanilaG commented on TrackerYP/Emoji.swift:
```diff
@@ -0,0 +1,7 @@
+import Foundation
+ 
+let emojiArray = [
+    "🙂", "😻", "🌺", "🐶", "❤️", "😱",
+    "😇", "😡", "🥶", "🤔", "🙌", "🍔",
+    "🥦", "🏓", "🥇", "🎸", "🏝", "😪",
+]
```
Супер, что вынес все эмоджи и цвета в одно место. Это сильно упростит работу, если их нужно будет поменять)

Commit:
1) Внес исправления в метод configureNavigationBar()
2) Удалил настройки для datePicker и addButton для view и из списка на добавление на view
3) Обернул TrackersViewController в UINavigationController на этапе добавления в иерархию
P.S. Данила, если у меня снова будет проблема с NavBar напишите, пожалуйста, в комментариях на GitHub, мол, ты дебил и ничего не исправил, чтобы мне не договариваться о еще 1 попытке сдачи. Спасибо.

DanilaG commented:
Договариваться о еще 1 попытке сдачи не придётся) Все супер, работу принемаю)

YanikMax commented:
> Договариваться о еще 1 попытке сдачи не придётся) Все супер, работу принемаю)

Спасибо :)
